### PR TITLE
add kafka resources operations

### PR DESCRIPTION
**Pull request by digna-ionos**
## Checklist

- [ ] PR name added as appropriate (e.g. `feat:`/`fix:`/`doc:`/`test:`/`refactor:`)
- [ ] Tests added or updated
- [ ] Documentation updated
- [ ] Sonar Cloud Scan
- [ ] Changelog updated and version incremented (label: upcoming release)
- [ ] Github Issue linked if any
- [ ] Jira task updated

### CODE REVIEW

**Comment by avirtopeanu-ionos** [comment]
nice work :+1: let's add some bats tests for this

File: commands/kafka/cluster/list.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		},
+	)
+
+	cmd.AddStringFlag(constants.FlagFilterName, "", "", "Filter used to fetch only the records that contain specified name.")
+	cmd.AddSetFlag(constants.FlagFilterState, "", "", []string{"AVAILABLE", "BUSY", "DEPLOYING", "UPDATING", "FAILED_UPDATING", "FAILED", "DESTROYING"}, "Filter used to fetch only the records that contain specified state.")
+	_ = cmd.Command.RegisterFlagCompletionFunc(constants.FlagFilterState, func(c *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		return []string{"AVAILABLE", "BUSY", "DEPLOYING", "UPDATING", "FAILED_UPDATING", "FAILED", "DESTROYING"}, cobra.ShellCompDirectiveNoFileComp
+	})
+
+	cmd.AddStringFlag(constants.FlagLocation, "", "", "The datacenter location", core.RequiredFlagOption())
```
Comment (avirtopeanu-ionos) [inline commands/kafka/cluster/list.go:48]: If I want to list clusters across all locations, I have to make multiple `list`s with multiple locations, right?

File: commands/kafka/cluster/list.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+				return listClusters(c)
+			},
+			InitClient: true,
+		},
+	)
+
+	cmd.AddStringFlag(constants.FlagFilterName, "", "", "Filter used to fetch only the records that contain specified name.")
+	cmd.AddSetFlag(constants.FlagFilterState, "", "", []string{"AVAILABLE", "BUSY", "DEPLOYING", "UPDATING", "FAILED_UPDATING", "FAILED", "DESTROYING"}, "Filter used to fetch only the records that contain specified state.")
+	_ = cmd.Command.RegisterFlagCompletionFunc(constants.FlagFilterState, func(c *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		return []string{"AVAILABLE", "BUSY", "DEPLOYING", "UPDATING", "FAILED_UPDATING", "FAILED", "DESTROYING"}, cobra.ShellCompDirectiveNoFileComp
```
Comment (avirtopeanu-ionos) [inline commands/kafka/cluster/list.go:45]: You don't need to add completions for `Set` type flags, they're added automatically

File: commands/kafka/cluster/create.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+func addClusterCreateFlags(cmd *core.Command) *core.Command {
+	cmd.AddStringFlag(constants.FlagName, "", "", "The name of the kafka cluster", core.RequiredFlagOption())
+	cmd.AddStringFlag(constants.FlagVersion, "", "", "The version of the kafka cluster", core.RequiredFlagOption())
+	cmd.AddSetFlag(constants.FlagSize, "", "", []string{"XS", "S", "M", "L", "XL"}, "The size of the kafka cluster", core.RequiredFlagOption())
+	_ = cmd.Command.RegisterFlagCompletionFunc(constants.FlagSize, func(c *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		return []string{"XS", "S", "M", "L", "XL"}, cobra.ShellCompDirectiveNoFileComp
+	})
+
+	cmd.AddStringFlag(constants.FlagLocation, "", "", "The location of the kafka cluster", core.RequiredFlagOption())
```
Comment (avirtopeanu-ionos) [inline commands/kafka/cluster/create.go:80]: Could we add this as a persistent flag on the `kafka` command?

Also, some users can't use completions due to the nature of their environment so we should add the possible values to the helptext of the flag

` slices.Collect(maps.Keys(locationToURL))`

File: commands/kafka/cluster/cluster.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+func Command() *core.Command {
+	cmd := &core.Command{
+		Command: &cobra.Command{
+			Use:              "cluster",
+			Short:            "The sub-commands of 'ionosctl kafka cluster' allow you to manage kafka clusters",
+			Aliases:          []string{"cl"},
+			TraverseChildren: true,
+		},
```
Comment (avirtopeanu-ionos) [inline commands/kafka/cluster/cluster.go:22]:
```suggestion
	PersistentPreRun: func(cmd *cobra.Command, args []string) {
		location, _ := cmd.Flags().GetString(constants.FlagLocation)
		changeLocation(client.Must().KafkaClient, location)
	},
```

after moving Lcation as a  PersistentFlag on this command

File: .goreleaser.yml [hunk 5]
```diff
@@ -70,6 +70,14 @@ brews:
     install: |-
       bin.install "ionosctl"
 
+snapcrafts:
```
Comment (avirtopeanu-ionos) [inline .goreleaser.yml:73]: is this meant to be here?

Comment (glimberea) [reply .goreleaser.yml:73]: nope, didn't notice it still being there, will remove it

File: test/bats/kafka/kafka.bats [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    datacenter_id=$(find_or_create_resource \
+        "ionosctl datacenter list -M 1 -F location=de/fra,state=available -o json 2> /dev/null | jq -r '.items[] | .id'" \
+        "ionosctl datacenter create --name \"CLI-Test-$(randStr 8)\" --location de/fra -o json 2> /dev/null | jq -r '.id'")
+    [ -n "$datacenter_id" ] || fail "datacenter_id is empty"
+    assert_regex "$datacenter_id" "$uuid_v4_regex"
+    echo "$datacenter_id" > /tmp/bats_test/datacenter_id
+    retry_until "ionosctl datacenter get --datacenter-id ${datacenter_id} -o json 2> /dev/null | jq -r '.metadata.state'" \
+        "[[ \$output == \"AVAILABLE\" ]]" 10 60
+
+    lan_id=$(find_or_create_resource \
```
Comment (avirtopeanu-ionos) [inline test/bats/kafka/kafka.bats:26]: This is more of a comment, and a personal anecdote, I found that using `find_or_create_resource` actually makes the test much more unpredictable, hard to debug, hard to read, hard to maintain etc.

You can obviously continue using it, but I have stopped using it for above reasons.

File: commands/cloudapi-v6/completer/ids.go [hunk 7]
```diff
@@ -52,7 +52,7 @@ func AttachedCdromsIds(datacenterId, serverId string) []string {
 	return attachedCdromsIds
 }
 
-func DataCentersIds() []string {
+func DataCentersIds(filters ...func(datacenter ionoscloud.Datacenter) bool) []string {
```
Comment (avirtopeanu-ionos) [inline commands/cloudapi-v6/completer/ids.go:55]: Nice addition!

File: test/bats/kafka/kafka.bats (hunk 6 above)
Comment (glimberea) [reply test/bats/kafka/kafka.bats:26]: Hmmm, I see, maybe we could look into removing it altogether from tests, in another PR of course. I will remove it from these tests for now.

**Review by avirtopeanu-ionos — APPROVED** [review]

**Pull request merged**